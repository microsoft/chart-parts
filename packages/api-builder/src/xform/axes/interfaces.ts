/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { SceneFrame } from '../context/SceneFrame'
import { SGMark, Axis, Scale, TickValue } from '@chart-parts/interfaces'

export interface AxisSpace {
	top: number
	right: number
	bottom: number
	left: number
}

/**
 * A context object for generating axis components
 */
export interface AxisContext {
	axis: Axis
	thickness: number
	scale: Scale<any, any>
	range: [number, number]
	rangeProperty: string
	rangeEndProperty: string
	crossProperty: string
	crossEndProperty: string
	horizontal: boolean
	topOrLeft: boolean
	frame: SceneFrame

	// Domain Data
	domainCross: number
	domainMinRange: number
	domainMaxRange: number

	// Tick Data
	ticks: PositionedTickValue[]
	tickCrossStart: number
	tickCrossEnd: number

	// Labels
	labelFormatter: (input: any) => string
}

export interface PositionedTickValue extends TickValue {
	position: number
}

/**
 * Interface for axis components such as ticks and labels
 */
export interface AxisComponent {
	/**
	 * Constructs the AxisContext object, which is used to render all axis components. Any
	 * pieces of context which this component generates or knows about should be populated here.
	 *
	 * @param context The intermediate context-object, which may be partially defined at this point
	 */
	createContext(context: Partial<AxisContext>): Partial<AxisContext>

	/**
	 * Determines whether the axis configuration results in this component emitting any information.
	 *
	 * @param context The axis context object
	 */
	isScenegraphElementGenerated(context: AxisContext): boolean

	/**
	 * Creates the scenegraph items for this axis component
	 *
	 * @param context The axis context object
	 */
	createScenegraphElement(context: AxisContext): SGMark<any>
}
