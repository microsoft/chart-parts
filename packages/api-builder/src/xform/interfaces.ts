/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { SGMark, SGItem, Channels } from '@chart-parts/interfaces'

/**
 * Interface for the result of scene-generation
 * @category Transformation
 */

export interface GeneratedScenegraph {
	/**
	 * The root mark of the resultant scene
	 */
	root: SGMark<SGItem>

	/**
	 * The event channels of this scene, a key of channel name to event handlers
	 */
	channelHandlers: Channels
}
