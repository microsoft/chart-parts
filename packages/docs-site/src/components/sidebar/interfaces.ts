/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

export interface SidebarItem {
	path: string
	title: string
	area: string
	order: number
}

export interface TreeNode {
	item: SidebarItem
	pathKey: string
	order?: number
	children: { [key: string]: TreeNode }
}
