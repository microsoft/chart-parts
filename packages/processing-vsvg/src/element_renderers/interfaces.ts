/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import {
	MarkConverter,
	VSvgNode,
	VSvgTransformType,
} from '@chart-parts/interfaces'

export interface VSvgMarkOutput {
	nodes: VSvgNode[]
	defs?: VSvgNode[]
}

export interface VSvgRenderContext {
	nextId(): string
}

export type VSvgMarkConverter = MarkConverter<VSvgMarkOutput, VSvgRenderContext>

export function translate(x: number, y: number) {
	return { type: VSvgTransformType.translate, value: [x, y] }
}

export function rotate(angle: number) {
	return { type: VSvgTransformType.rotate, value: angle }
}
