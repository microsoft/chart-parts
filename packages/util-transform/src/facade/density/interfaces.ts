/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { DatasetManager } from '../dataset'
import { DatasetTransform } from '../interfaces'

export interface DistributionBuilder {
	build(ds: DatasetManager): any
}

/**
 * The density transform generates a new data stream of uniformly-spaced
 * samples drawn from a one-dimensional probability density function (pdf) or
 * cumulative distribution function (cdf). This transform is useful for
 * representing probability distributions and generating continuous
 * distributions from discrete samples using kernel density estimation.
 */
export interface DensityBuilder extends DatasetTransform {
	/**
	 * Required. An object describing the distribution type and parameters.
	 * See the distribution reference for more.
	 *
	 * @param dist
	 */
	distribution(dist: DistributionBuilder): DensityBuilder

	/**
	 * A [min, max] domain from which to sample the distribution. This argument
	 * is required in most cases, but can be omitted in the case of
	 * distributions (namely, kde) that can deduce their own extent.
	 *
	 * @param start
	 * @param end
	 */
	extent(start: number, end: number): DensityBuilder

	/**
	 * The type of distribution to generate.
	 * @param value
	 */
	method(value: DensityMethod): DensityBuilder

	/**
	 * The number of uniformly spaced steps to take along the extent domain
	 * (default 100). A total of steps + 1 uniformly-spaced samples are drawn
	 * from the distribution.
	 *
	 * @param value
	 */
	steps(value: number): DensityBuilder

	/**
	 * The output fields for the sample value and associated probability.
	 * The default is ["value", "density"].
	 *
	 * @param valueString
	 * @param probabilityField
	 */
	as(valueString: string, probabilityField: string): DensityBuilder
}

export enum DensityMethod {
	/**
	 * Probability Density Funciton
	 */
	pdf = 'pdf',

	/**
	 * Cumulative Distribution Function
	 */
	cdf = 'cdf',
}
