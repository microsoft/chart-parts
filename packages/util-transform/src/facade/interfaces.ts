/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { DatasetManager } from './dataset'

export interface DatasetTransform {
	/**
	 * Build the transformation
	 * @param df The vega dataflow instance
	 * @param from The vega-dataflow node that is the parent of this node
	 * @param ds The chart-parts dataset
	 */
	build(df: any, from: any, ds: DatasetManager): any
}
