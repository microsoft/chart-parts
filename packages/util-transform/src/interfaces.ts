/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

export type FieldAccessor = string // | ((row: any) => number)

export enum TextCaseTransform {
	lower = 'lower',
	upper = 'upper',
	mixed = 'mixed',
}

export enum CompareOrder {
	ascending = 'asc',
	descending = 'desc',
}

export interface Compare {
	field: FieldAccessor
	order?: CompareOrder
}

export enum Offset {
	zero = 'zero',
	center = 'center',
	normalize = 'normalize',
}

export type CompareFunction<T> = (a: T, b: T) => number
export type EqualityFunction<T> = (a: T, b: T) => boolean
export type GetKeyFunction<T> = (item: T) => any
export type Predicate<T> = (item: T) => boolean
export type Transformer<T, K> = (input: T) => K

export type MaybeNumber = number | undefined
