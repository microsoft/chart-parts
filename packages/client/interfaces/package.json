{
	"name": "@chart-parts/interfaces",
	"description": "Common interfaces for the visualization system, including marks, mark specifications, scenegraphs, and virtual dom definitions",
	"version": "0.1.3",
	"main": "dist/cjs/index.js",
	"module": "dist/esm/index.js",
	"types": "dist/types/index.d.ts",
	"license": "MIT",
	"scripts": {
		"clean": "essex clean",
		"build": "essex build",
		"start": "essex watch"
	},
	"publishConfig": {
		"access": "public"
	},
	"devDependencies": {
		"@types/node": "^14.0.11"
	}
}
