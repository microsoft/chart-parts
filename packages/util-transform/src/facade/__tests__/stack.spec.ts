/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { stack, dataset } from '..'
import { Offset, CompareOrder } from '../../interfaces'

const data = [
	{ x: 0, y: 28, c: 0 },
	{ x: 0, y: 55, c: 1 },
	{ x: 1, y: 43, c: 0 },
	{ x: 1, y: 91, c: 1 },
	{ x: 2, y: 81, c: 0 },
	{ x: 2, y: 53, c: 1 },
	{ x: 3, y: 19, c: 0 },
	{ x: 3, y: 87, c: 1 },
	{ x: 4, y: 52, c: 0 },
	{ x: 4, y: 48, c: 1 },
	{ x: 5, y: 24, c: 0 },
	{ x: 5, y: 49, c: 1 },
	{ x: 6, y: 87, c: 0 },
	{ x: 6, y: 66, c: 1 },
	{ x: 7, y: 17, c: 0 },
	{ x: 7, y: 27, c: 1 },
	{ x: 8, y: 68, c: 0 },
	{ x: 8, y: 16, c: 1 },
	{ x: 9, y: 49, c: 0 },
	{ x: 9, y: 15, c: 1 },
]

describe('The stack transformation', () => {
	it('can stack values', () => {
		const ds = dataset()
			.addTable('data', data, stack('y'))
			.addTable('data-grouped', data, stack('y').groupBy('x'))
			.addTable(
				'data-grouped-sorted-and-offset',
				data,
				stack('y')
					.groupBy('x')
					.sort({ field: 'c', order: CompareOrder.descending })
					.offset(Offset.center),
			)

		// The result should not be the same instance (i.e. preserve the input for functional style)
		expect(ds.getTable('data')).toMatchSnapshot()
		expect(ds.getTable('data-grouped')).toMatchSnapshot()
		expect(ds.getTable('data-grouped-sorted-and-offset')).toMatchSnapshot()
	})
})
