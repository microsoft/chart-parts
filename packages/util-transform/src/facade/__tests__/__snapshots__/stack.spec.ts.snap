// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The stack transformation can stack values 1`] = `
Array [
  Object {
    "c": 0,
    "x": 0,
    "y": 28,
    "y0": 90,
    "y1": 118,
    Symbol(vega_id): 1,
  },
  Object {
    "c": 1,
    "x": 0,
    "y": 55,
    "y0": 35,
    "y1": 90,
    Symbol(vega_id): 2,
  },
  Object {
    "c": 0,
    "x": 1,
    "y": 43,
    "y0": 100.5,
    "y1": 143.5,
    Symbol(vega_id): 3,
  },
  Object {
    "c": 1,
    "x": 1,
    "y": 91,
    "y0": 9.5,
    "y1": 100.5,
    Symbol(vega_id): 4,
  },
  Object {
    "c": 0,
    "x": 2,
    "y": 81,
    "y0": 62.5,
    "y1": 143.5,
    Symbol(vega_id): 5,
  },
  Object {
    "c": 1,
    "x": 2,
    "y": 53,
    "y0": 9.5,
    "y1": 62.5,
    Symbol(vega_id): 6,
  },
  Object {
    "c": 0,
    "x": 3,
    "y": 19,
    "y0": 110.5,
    "y1": 129.5,
    Symbol(vega_id): 7,
  },
  Object {
    "c": 1,
    "x": 3,
    "y": 87,
    "y0": 23.5,
    "y1": 110.5,
    Symbol(vega_id): 8,
  },
  Object {
    "c": 0,
    "x": 4,
    "y": 52,
    "y0": 74.5,
    "y1": 126.5,
    Symbol(vega_id): 9,
  },
  Object {
    "c": 1,
    "x": 4,
    "y": 48,
    "y0": 26.5,
    "y1": 74.5,
    Symbol(vega_id): 10,
  },
  Object {
    "c": 0,
    "x": 5,
    "y": 24,
    "y0": 89,
    "y1": 113,
    Symbol(vega_id): 11,
  },
  Object {
    "c": 1,
    "x": 5,
    "y": 49,
    "y0": 40,
    "y1": 89,
    Symbol(vega_id): 12,
  },
  Object {
    "c": 0,
    "x": 6,
    "y": 87,
    "y0": 66,
    "y1": 153,
    Symbol(vega_id): 13,
  },
  Object {
    "c": 1,
    "x": 6,
    "y": 66,
    "y0": 0,
    "y1": 66,
    Symbol(vega_id): 14,
  },
  Object {
    "c": 0,
    "x": 7,
    "y": 17,
    "y0": 81.5,
    "y1": 98.5,
    Symbol(vega_id): 15,
  },
  Object {
    "c": 1,
    "x": 7,
    "y": 27,
    "y0": 54.5,
    "y1": 81.5,
    Symbol(vega_id): 16,
  },
  Object {
    "c": 0,
    "x": 8,
    "y": 68,
    "y0": 50.5,
    "y1": 118.5,
    Symbol(vega_id): 17,
  },
  Object {
    "c": 1,
    "x": 8,
    "y": 16,
    "y0": 34.5,
    "y1": 50.5,
    Symbol(vega_id): 18,
  },
  Object {
    "c": 0,
    "x": 9,
    "y": 49,
    "y0": 59.5,
    "y1": 108.5,
    Symbol(vega_id): 19,
  },
  Object {
    "c": 1,
    "x": 9,
    "y": 15,
    "y0": 44.5,
    "y1": 59.5,
    Symbol(vega_id): 20,
  },
]
`;

exports[`The stack transformation can stack values 2`] = `
Array [
  Object {
    "c": 0,
    "x": 0,
    "y": 28,
    "y0": 90,
    "y1": 118,
    Symbol(vega_id): 1,
  },
  Object {
    "c": 1,
    "x": 0,
    "y": 55,
    "y0": 35,
    "y1": 90,
    Symbol(vega_id): 2,
  },
  Object {
    "c": 0,
    "x": 1,
    "y": 43,
    "y0": 100.5,
    "y1": 143.5,
    Symbol(vega_id): 3,
  },
  Object {
    "c": 1,
    "x": 1,
    "y": 91,
    "y0": 9.5,
    "y1": 100.5,
    Symbol(vega_id): 4,
  },
  Object {
    "c": 0,
    "x": 2,
    "y": 81,
    "y0": 62.5,
    "y1": 143.5,
    Symbol(vega_id): 5,
  },
  Object {
    "c": 1,
    "x": 2,
    "y": 53,
    "y0": 9.5,
    "y1": 62.5,
    Symbol(vega_id): 6,
  },
  Object {
    "c": 0,
    "x": 3,
    "y": 19,
    "y0": 110.5,
    "y1": 129.5,
    Symbol(vega_id): 7,
  },
  Object {
    "c": 1,
    "x": 3,
    "y": 87,
    "y0": 23.5,
    "y1": 110.5,
    Symbol(vega_id): 8,
  },
  Object {
    "c": 0,
    "x": 4,
    "y": 52,
    "y0": 74.5,
    "y1": 126.5,
    Symbol(vega_id): 9,
  },
  Object {
    "c": 1,
    "x": 4,
    "y": 48,
    "y0": 26.5,
    "y1": 74.5,
    Symbol(vega_id): 10,
  },
  Object {
    "c": 0,
    "x": 5,
    "y": 24,
    "y0": 89,
    "y1": 113,
    Symbol(vega_id): 11,
  },
  Object {
    "c": 1,
    "x": 5,
    "y": 49,
    "y0": 40,
    "y1": 89,
    Symbol(vega_id): 12,
  },
  Object {
    "c": 0,
    "x": 6,
    "y": 87,
    "y0": 66,
    "y1": 153,
    Symbol(vega_id): 13,
  },
  Object {
    "c": 1,
    "x": 6,
    "y": 66,
    "y0": 0,
    "y1": 66,
    Symbol(vega_id): 14,
  },
  Object {
    "c": 0,
    "x": 7,
    "y": 17,
    "y0": 81.5,
    "y1": 98.5,
    Symbol(vega_id): 15,
  },
  Object {
    "c": 1,
    "x": 7,
    "y": 27,
    "y0": 54.5,
    "y1": 81.5,
    Symbol(vega_id): 16,
  },
  Object {
    "c": 0,
    "x": 8,
    "y": 68,
    "y0": 50.5,
    "y1": 118.5,
    Symbol(vega_id): 17,
  },
  Object {
    "c": 1,
    "x": 8,
    "y": 16,
    "y0": 34.5,
    "y1": 50.5,
    Symbol(vega_id): 18,
  },
  Object {
    "c": 0,
    "x": 9,
    "y": 49,
    "y0": 59.5,
    "y1": 108.5,
    Symbol(vega_id): 19,
  },
  Object {
    "c": 1,
    "x": 9,
    "y": 15,
    "y0": 44.5,
    "y1": 59.5,
    Symbol(vega_id): 20,
  },
]
`;

exports[`The stack transformation can stack values 3`] = `
Array [
  Object {
    "c": 0,
    "x": 0,
    "y": 28,
    "y0": 90,
    "y1": 118,
    Symbol(vega_id): 1,
  },
  Object {
    "c": 1,
    "x": 0,
    "y": 55,
    "y0": 35,
    "y1": 90,
    Symbol(vega_id): 2,
  },
  Object {
    "c": 0,
    "x": 1,
    "y": 43,
    "y0": 100.5,
    "y1": 143.5,
    Symbol(vega_id): 3,
  },
  Object {
    "c": 1,
    "x": 1,
    "y": 91,
    "y0": 9.5,
    "y1": 100.5,
    Symbol(vega_id): 4,
  },
  Object {
    "c": 0,
    "x": 2,
    "y": 81,
    "y0": 62.5,
    "y1": 143.5,
    Symbol(vega_id): 5,
  },
  Object {
    "c": 1,
    "x": 2,
    "y": 53,
    "y0": 9.5,
    "y1": 62.5,
    Symbol(vega_id): 6,
  },
  Object {
    "c": 0,
    "x": 3,
    "y": 19,
    "y0": 110.5,
    "y1": 129.5,
    Symbol(vega_id): 7,
  },
  Object {
    "c": 1,
    "x": 3,
    "y": 87,
    "y0": 23.5,
    "y1": 110.5,
    Symbol(vega_id): 8,
  },
  Object {
    "c": 0,
    "x": 4,
    "y": 52,
    "y0": 74.5,
    "y1": 126.5,
    Symbol(vega_id): 9,
  },
  Object {
    "c": 1,
    "x": 4,
    "y": 48,
    "y0": 26.5,
    "y1": 74.5,
    Symbol(vega_id): 10,
  },
  Object {
    "c": 0,
    "x": 5,
    "y": 24,
    "y0": 89,
    "y1": 113,
    Symbol(vega_id): 11,
  },
  Object {
    "c": 1,
    "x": 5,
    "y": 49,
    "y0": 40,
    "y1": 89,
    Symbol(vega_id): 12,
  },
  Object {
    "c": 0,
    "x": 6,
    "y": 87,
    "y0": 66,
    "y1": 153,
    Symbol(vega_id): 13,
  },
  Object {
    "c": 1,
    "x": 6,
    "y": 66,
    "y0": 0,
    "y1": 66,
    Symbol(vega_id): 14,
  },
  Object {
    "c": 0,
    "x": 7,
    "y": 17,
    "y0": 81.5,
    "y1": 98.5,
    Symbol(vega_id): 15,
  },
  Object {
    "c": 1,
    "x": 7,
    "y": 27,
    "y0": 54.5,
    "y1": 81.5,
    Symbol(vega_id): 16,
  },
  Object {
    "c": 0,
    "x": 8,
    "y": 68,
    "y0": 50.5,
    "y1": 118.5,
    Symbol(vega_id): 17,
  },
  Object {
    "c": 1,
    "x": 8,
    "y": 16,
    "y0": 34.5,
    "y1": 50.5,
    Symbol(vega_id): 18,
  },
  Object {
    "c": 0,
    "x": 9,
    "y": 49,
    "y0": 59.5,
    "y1": 108.5,
    Symbol(vega_id): 19,
  },
  Object {
    "c": 1,
    "x": 9,
    "y": 15,
    "y0": 44.5,
    "y1": 59.5,
    Symbol(vega_id): 20,
  },
]
`;
